sdks/java/core/src/main/java/org/apache/beam/sdk/util/VarInt.java

```diff
@@ -42,13 +42,61 @@ public static void encode(int v, OutputStream stream) throws IOException {
 
   /** Encodes the given value onto the stream. */
   public static void encode(long v, OutputStream stream) throws IOException {
-    do {
-      // Encode next 7 bits + terminator bit
-      long bits = v & 0x7F;
-      v >>>= 7;
-      byte b = (byte) (bits | ((v != 0) ? 0x80 : 0));
-      stream.write(b);
-    } while (v != 0);
+    if ((v & ~0x7F) == 0) {
```

Comment:
I happen to be browsing the history, and this sort of change really needs a comment so that someone landing on it knows the details of the decision, such as a summary or even a link to the numbers in this PR.

Reply:
(or else they might think "meh I'll change this to simplify")